sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <azure/core/http/policies/policy.hpp>
+
+namespace Azure { namespace Storage { namespace _internal {
+
+  class StorageBearerTokenAuthenticationPolicy final
+      : public Core::Http::Policies::_internal::BearerTokenAuthenticationPolicy {
+  public:
+    /**
+     * @brief Construct a Storage Bearer Token challenge authentication policy.
+     *
+     * @param credential An #Azure::Core::TokenCredential to use with this policy.
+     * @param tokenRequestContext A context to get the token in.
+     * @param enableTenantDiscovery Enables tenant discovery through the authorization challenge.
+     */
+    explicit StorageBearerTokenAuthenticationPolicy(
+        std::shared_ptr<const Azure::Core::Credentials::TokenCredential> credential,
+        Azure::Core::Credentials::TokenRequestContext tokenRequestContext,
+        bool enableTenantDiscovery)
+        : BearerTokenAuthenticationPolicy(std::move(credential), tokenRequestContext),
+          m_Scopes(tokenRequestContext.Scopes), m_TenantId(tokenRequestContext.TenantId),
+          m_EnableTenantDiscovery(enableTenantDiscovery)
+    {
+    }
+
+    ~StorageBearerTokenAuthenticationPolicy() override {}
+
+    std::unique_ptr<HttpPolicy> Clone() const override
+    {
+      return std::unique_ptr<HttpPolicy>(new StorageBearerTokenAuthenticationPolicy(*this));
+    }
+
+  private:
+    std::vector<std::string> m_Scopes;
+    mutable std::string m_TenantId;
+    mutable std::mutex m_TenantIdMutex;
```

Comment:
You can do better than the plain `std::mutex`  - use `std::shared_timed_mutex` (unlike `std::shared_mutex`, it is available in C++14), and then you can do `std::shared_lock` for reading and `std::unique_lock` for writing.
For examples, one of the places we're doing this, is in core logging functionality: https://github.com/Azure/azure-sdk-for-cpp/blob/0b59e5a94c3551cf550196db1ee7175972e69d10/sdk/core/azure-core/src/logger.cpp

Reply:
Thanks for your suggestion. I will use `std::shared_timed_mutex` and improve this code.